Add virtual machines data source

jake2184 opened the pull request:
Add 'virtual machines' as a data source.

I'm using this provider to extract information from Netbox, and use that to configure a RHEV/oVirt virtualisation platform - this was a necessary addition.

If someone provides docs about your test data set, I'm happy to add tests for this PR.

fbreckle commented on netbox/data_source_netbox_virtual_machines.go:
```diff
@@ -0,0 +1,232 @@
+			case "name":
+				var nameString = v.(string)
+				params.Name = &nameString
+			case "region":
+				var siteString = v.(string)
+				params.Site = &siteString
```
This should probably say regionString etc.

jake2184 replied:
Oops. Fixed

jake2184 commented:
To work with ConfigContext (and Local Context) , this is pending https://github.com/fbreckle/go-netbox/pull/2

fbreckle commented:
oh and note that `testAccNetboxVirtualMachineFullDependencies` in `resource_netbox_virtual_machine_test.go` is an example on how bigger dependency chains are created at the moment. It probably makes sense to use that function in the tests for your `virtual_machines` data source.

jake2184 commented:
I've:
- Added a `name_regex` parameter
- Added testing
- Implemented searching via a `filter` block functionality similar to AWS.

I was using search as I was modelling code on `terraform-provider-ovirt`, as I'm new to this and that's the code base I'm used to. However it's not necessarily a good guide - it has some opaque requirements for testing, hence my question about data set.

fbreckle commented on netbox/data_source_netbox_virtual_machines_test.go:
```diff
@@ -0,0 +1,111 @@
+data "netbox_virtual_machines" "test" {
+  filter {
+    name  = "cluster_id"
+	value = netbox_cluster.test.id
```
can you use spaces here

fbreckle commented on netbox/data_source_netbox_virtual_machines_test.go:
```diff
@@ -0,0 +1,111 @@
+func testAccNetboxVirtualMachineDataSourceDependencies(testName string) string {
+	return testAccNetboxVirtualMachineFullDependencies(testName) + fmt.Sprintf(`
+resource "netbox_virtual_machine" "test0" {
+	name = "%[1]s_0"
+	cluster_id = netbox_cluster.test.id
+	comments = "thisisacomment"
+	memory_mb = 1024
+	disk_size_gb = 256
+	tenant_id = netbox_tenant.test.id
+	role_id = netbox_device_role.test.id
+	platform_id = netbox_platform.test.id
+	vcpus = 4
+}
```
can you use spaces here. Indent tf code with 2 spaces and go code with tabs. At least thats the current standard and it should be consistent.

jake2184 replied:
hmm, that was copied and pasted from another file. Probably my editor auto changing. Will fixup

fbreckle commented on netbox/data_source_netbox_virtual_machines_test.go:
```diff
@@ -0,0 +1,111 @@
+					resource.TestCheckResourceAttrPair("data.netbox_virtual_machines.test", "vms.1.name", "netbox_virtual_machine.test1", "name"),
+					resource.TestCheckResourceAttrPair("data.netbox_virtual_machines.test", "vms.2.name", "netbox_virtual_machine.test2", "name"),
+					resource.TestCheckResourceAttrPair("data.netbox_virtual_machines.test", "vms.3.name", "netbox_virtual_machine.test3", "name"),
+
+
```
Extra newlines?

fbreckle commented:
This is looking really great. I'm somewhat concerned about the order that netbox API returns the VMs in. Apparently it returns them in the order created, which is good enough for the tests. A quick google search says that explicit sorting is not an option anyway.

I'll merge once the formatting is clean. You should also run `go fmt`, it creates a diff in `data_source_netbox_virtual_machines.go` for me.

jake2184 commented:
Typically you wouldn't rely on the order of a returned list, or at least I would avoid it for that reason. 

For example, in my use case I'm using `for_each` on the returned list, converting it to a map.
